### ISSUE TITLE
Support Rabbitmq´s MQTT plugin

### ISSUE THREAD

**Issue by petersilva** [MetPX/sarracenia#227]
Sibling enhancement to #128 rather than have Sarracenia communicate at the application level to MQTT brokers, just take advantage of the mqtt plugin built-in to rabbitmq.
Could have an option like so:

```

rabbitmqtt_post_broker amqp://feeder@localhost
rabbitmqtt_post_exchange amq.topic

```

Given those settings, Sarracenia would initiate an AMQP connection to the given rabbitmq broker, and publish to the rabbitmqtt_post_exchange.  the topic would be
prefixed the value of the *post_exchange* 

So this should be easily done with a low-level tweak, perhaps in sr_message, or sr_amqp.

**Comment by petersilva** [comment]
Haven´t mentioned it yet, but obviously, the target rabbitmq broker must have mqtt plugin enabled and configured to match the settings above.  Then something published on amqp as:

```
Exchange: xpublic,   topic:  20190616.GTS.CWAO.SA ...

will show up in the MQTT exchange as topic:   xpublic/20190616/GTS/CWAO/SA ...
```

**Comment by petersilva** [comment]
Once we can publish for MQTT subscribers, a little additional work would be needed to enable a foreign (non-sarracenia) MQTT client could publish to the hierarchy on the rabbitmq broker, and sarracenia could subscribe via amqp without modification...

I think all we need is:

`
rabbitmqtt_exchange  amq.topic
`

and the sarracenia consumer, will know to take the first topic, and pretend that is an exchange, reversing the transformation above.

**Comment by petersilva** [comment]
so with three options, and some small tweaks in message parsing, we could get likely one model of mqtt interop working.

**Comment by petersilva** [comment]
Perhaps this is sufficiently confusing that we should write the documentation first?

**Comment by benlapETS** [comment]
Note1: All I have for now about this issue are questions :
- How are Sarra users flexible to learn new configurations ? 
- Is it too much for the sole goal of supporting MQTT ? 
- Will it be confusing even with good documentation ?
In brief, I would like to get some feedback from Sarra users (mostly analyst?)  about their feeling on providing 2 solution (#128 and #227) to one problem adding multiple new configuration ?
Note2: I want to provide a diagram of an overview of this current use case, I will work on this today.

**Comment by petersilva** [comment]
review of offline chat with @benlapETS :

* Sarracenia makes a point of not adding any abstraction of the underlying services.  It tries to use the AMQP broker as-is, the apache web server as-is, ssh server as-is, so the configuration of those things are documented by the respective tools.  

* The rabbitmq mqtt plugin is a very specific technology, and building an abstraction around it to hide the implementation is likely to be a lot more difficult to work with going forward. The plan outlined above is to have the skinniest of wrapping explicitly around the rabbitmq plugin, to allow sarracenia components to use it with minimal changes.

* the analysts we could consult with effectively are the two of us, as no-one else in our team has much exposure to mqtt brokers or the plugin for rabbitmq.  It is very difficult to have productive consultations without working code.

* Probably the best way to consult with the analysts is to write the code and document it as EXPERIMENTAL.  THEN we can have analysts kick the tires, and provide feedback.  This method makes sense, because this feature is likely pretty short to implement.  It's harder to do if the investment to get to basic function is much higher (as will likely be the case for #128)

* We could implement #128 the same way (tagged EXPERIMENTAL) which allows us to remove the functionality later if we conclude that it isn't a good path.  I think we will end up keeping both though.

**Comment by benlapETS** [comment]
- The right procedure would be to document it in the .rst files and in the appropriate man pages I guess, isnt it ?

**Comment by petersilva** [comment]
yes, the man pages for sure. I guess the setup instructions would go in the Admin.rst, and the something should indicate the features are experimental.

**Comment by petersilva** [comment]
thinking about it... I don't think we even need the rabbitmqtt_post_broker setting.  We just use the ordinary post_broker,  the only setting we need is rabbitmqtt_post_exchange... when set the already described transformations happen to the topic and exchange, and we just use the existing post_broker.

**Comment by benlapETS** [comment]
Note1. Here are the detailed use cases for #227:
1. Sarracenia user want to activate the MQTT Plugin in one RabbitMQ broker. He decides of the name of the mqtt topic exchange to use, the set of mqtt credentials, the uses of anonymous, prefetch, the security support, vhost mapping. The information for those needs is all available at [MQTT Plugin](https://www.rabbitmq.com/mqtt.html) page. The configuration is usally done statically once. Then the system will set the configuration that we already support with sr_audit like adding exchange, adding user.  

2. Sarracenia is MQTT Plugin ready and the Sarracenia user want to publish message through the RabbitMQ broker. He activate the rabbit mqtt post broker config through Sarracenia which will publish every message that it translate to MQTT msg prefixed with the amqp exchange that it would normally use.

3. Sarracenia is MQTT Plugin ready and the Sarracenia user want to consume messages through the RabbitMQ broker. He activate the rabbit mqtt broker through Sarracenia which will catch every mqtt message and translate it to normal amqp msg removing the exchange prefix on the topic. The msg will then be consume as a normal amqp msg.

4. A Partner User consume MQTT msgs from a Sarracenia node that has MQTT enabled with an external MQTT client.

Note2: Questions:
- Why would an analyst want to consume MQTT feed ?
   Nourredine told me that it is a possibility in an upgrade need from AMQP to MQTT standard but not in the near future. 
- If the end user doesnt use Sarracenia he wont get the data, because those MQTT client won't download files for him as Sarracenia does.  So, is it pointless to add this support or is it exactly what the end user wants, implement the download functionality itself ?
 To give users the ability (and the responsibility) to do anything they want with the data is exactly what we want and what some end users want.

**Comment by petersilva** [comment]
Note 2 questions... yes if all publishing is done from Sarracenia, and mqtt is only a consumer, then it doesn''t make much sense to consume from mqtt.  I'm not entirely sure of the permission model.  After we have done the first iteration, we may find different use cases, where we allow MQTT users to publish to the mqtt side of the broker, and we want to forward to others.

The other, much simpler/obvious use case is for loopback testing with rabbitmqtt_post_broker which would likely be a good thing to add as a flow_test. chain (add rabbitmqtt plugin on flow test, publish and then consume through it.)

We won't have a Sarracenia MQTT consumer until #128 is done, but if we were only worried about Sarracenia, then there is no need for mqtt in the first place.  The purpose of adding MQTT is to demonstrate interop with other implementations... we could try it with the  https://github.com/MetPX/wmo_mesh/ project as a sample consumer, but there are no existing implementations to interop with yet.  so...

The motivations are:

Build it and they will come - https://www.youtube.com/watch?v=5Ay5GqJwHF8 (providing an intial reference implementation for others to work with.)

Dogfooding - https://medium.com/@benbob/the-gospel-of-dogfooding-can-i-hear-an-amen-brother-1af4d82cf221 ( using it ourselves to prove it works. )

**Comment by petersilva** [comment]
oh, yeah... there is another motivation for MQTT support... the dialup data acquisition (DARTS) stuff will evolve over the next few years, and there seems to be industry weight behind MQTT, so it is possible that the auto stations will produce MQTT data natively.   Getting the data acquisition and exchange worlds closer together would be a good thing... but again, implementations have to exist to for people to work with.

**Comment by benlapETS** [comment]
Note1:
Functionnal requirements (new or preserved):
1. (preserved) Sarracenia must allow anonymous users.
2. (new) Sarracenia must allow MQTT users to publish to the MQTT side of the broker.
3. (new) Sarracenia must be able to relay a message received from MQTT to the appropriate AMQP in less than (1s).

Non functionnal requirements
1. Sarracenia must support MQTT model without any disruption to its own which is relying on AMQP standards

**Comment by benlapETS** [comment]
Note1: The actual configuration `exchange` and `post_exchange` are quite confusing, I propose to change it to: `from_exchange` and `to_exchange`

**Comment by petersilva** [comment]
*changing* will break a lot... Could add an alias though...

**Comment by benlapETS** [comment]
I don't intend to change right now, I'm just taking notes on my thoughts. I am kind of building a list of future proposition of enhancement if we make a big version change.

**Comment by benlapETS** [comment]
Note 1: My last thoughts on this issue will surely change some reasoning that has already been done on MQTT support with RabbitMQTT. Having already done some experimentations using the flow test, I realized I could create a flow of MQTT data by redirecting normal AMQP exchange flow to the designated MQTT exchange of one particular broker. To do that, I used this shovel configuration, without disrupting Sarracenia code:
```
#todo
```
So, this config enabled RabbitMQTT publishing and MQTT consuming through a third-party user running mosquitto_sub. Then, to consume through Sarracenia I used this subscriber config:
```
#todo
```
These config have already been integrated to the flow test, will need to document that and then discuss on the details, if this integration suites our needs for the RabbitMQ MQTT plugin support.

**Comment by benlapETS** [comment]
Note1: I wrote some [howto doc](https://github.com/MetPX/sarracenia/blob/master/doc/Admin.rst#mqtt-support) to enable MQTT with Sarracenia, this is a first draft, but I tested it on my other machine and it worked. The only problem is that I had to authorized anonymous to have .* read permission (couldn't restrict to only x.*public).

**Comment by petersilva** [comment]
the name used in the howto, _xmqtt_public_ exchange name doesn't match the required naming scheme ... which is why you needed to tweak permissions... the exchange name needs to start with xs_ and end in puiblic.  Try an exchange name: xs_mqtt_public, and I imagine anonymous will be able to read it without the permission tweak.  I think every sr_audit run will reset the permissions anyways... the recipe is likely very fragile as-is, it might even delete the exchange if run with the right options.

The permission model (which is based on standardized name scheme for exchanges) is described here:

https://github.com/MetPX/sarracenia/blob/master/doc/sr_subscribe.1.rst#naming-exchanges-and-queues

and a slightly older (and conflicting... ought to fix that ;-) here:

https://github.com/MetPX/sarracenia/blob/master/doc/Concepts.rst#security-considerations

**Comment by benlapETS** [comment]
Note1: The problem is not only the scheme, in order to allow Partner user subscribers (like mosquitto_sub) to a feed we should allow a different queue scheme which we don't control the format (for now) and which start with mqtt-subscription

**Comment by petersilva** [comment]
hmm... that is an issue... ok, if we need to change the permission mask, the solution is to modify sr_audit to enforce it always, and update documentation to suit. need to think about this for a while.

**Comment by petersilva** [comment]
ok... we discussed how you had proposed incorporating the topic prefix into the hierarchy at the end.  This has the effect of putting all mqtt traffic into a single hierarchy.  The reason for putting the exchange at the beginning of the tree was to be able to establish independent topic hierarchies (this things create data structures on mqtt brokers, much like exchanges on amqp ones.) so putting the exchange at the beginning spreads them out.  I worry that piling everything into a single v03.post hierarchy will have adverse effects.  Is this real?  dunno, thinking....

**Comment by petersilva** [comment]
the other worry is that there isn't any way to identify a posting (by the content itself, without knowing) that has the exchange as the first post, versus one that doesn't....  should we perhaps differentiate using v03.postx.   (if it has postx, then following subtopic is an exchange id?)

**Comment by petersilva** [comment]
released in 2.19.09
